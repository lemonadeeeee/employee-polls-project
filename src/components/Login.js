import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useLocation } from 'react-router-dom';
import { setUser } from '../features/auth/authSlice';
import { selectAllUsers } from '../features/users/usersSlice';
import styles from './Login.module.css';

function Login() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');
  
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const location = useLocation();
  const users = useSelector(selectAllUsers);
  
  const from = location.state?.from?.pathname || "/";

  const handleSubmit = (e) => {
    e.preventDefault();
    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );

    if (user) {
      dispatch(setUser(user));
      // navigate('/');
  	  navigate(from, { replace: true });
    } else {
      setError('Invalid username or password');
    }
  };

  return (
    <div className={styles.loginContainer}>
      <h2 className={styles.loginTitle}>Login</h2>
      {error && <p className={styles.error}>{error}</p>}
      <form onSubmit={handleSubmit} className={styles.loginForm}>
        <input
          type="text"
          placeholder="Username"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
          required
          className={styles.inputField}
        />
        <input
          type="password"
          placeholder="Password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          required
          className={styles.inputField}
        />
        <button type="submit" className={styles.loginButton}>Login</button>
      </form>
    </div>
  );
}

export default Login;