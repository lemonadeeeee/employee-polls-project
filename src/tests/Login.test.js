import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import Login from '../components/Login';

const mockStore = configureStore([]);

describe('Login Component', () => {
  it('should match snapshot', () => {
    const store = mockStore({ users: {} });
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  
  it('should have username field, password field, and submit button', () => {
    const store = mockStore({ users: {} });
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    expect(screen.getByPlaceholderText('Username')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /login/i })).toBeInTheDocument();
  });

  it('should update input fields and submit form', () => {
    const store = mockStore({
      users: {
        sarahedo: {
          id: 'sarahedo',
          password: 'password123',
          name: 'Sarah Edo',
          avatarURL: null,
          answers: {},
          questions: []
        }
      }
    });

    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const usernameInput = screen.getByPlaceholderText('Username');
    const passwordInput = screen.getByPlaceholderText('Password');
    const submitButton = screen.getByRole('button', { name: /login/i });

    fireEvent.change(usernameInput, { target: { value: 'sarahedo' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });

    expect(usernameInput).toHaveValue('sarahedo');
    expect(passwordInput).toHaveValue('password123');

    fireEvent.click(submitButton);

    // Check if the correct action was dispatched
    expect(store.getActions()).toContainEqual(expect.objectContaining({ 
      type: 'auth/setUser',
      payload: expect.objectContaining({ id: 'sarahedo' })
    }));
  });

  it('should display an error message for incorrect login', () => {
    const store = mockStore({ users: {} });
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    fireEvent.change(screen.getByPlaceholderText('Username'), { target: { value: 'wronguser' } });
    fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: 'wrongpass' } });
    fireEvent.click(screen.getByRole('button', { name: /login/i }));

    expect(screen.getByText('Invalid username or password')).toBeInTheDocument();
  });
});