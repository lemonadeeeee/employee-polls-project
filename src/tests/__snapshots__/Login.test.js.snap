// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="loginContainer"
  >
    <h2
      class="loginTitle"
    >
      Login
    </h2>
    <form
      class="loginForm"
    >
      <input
        class="inputField"
        placeholder="Username"
        required=""
        type="text"
        value=""
      />
      <input
        class="inputField"
        placeholder="Password"
        required=""
        type="password"
        value=""
      />
      <button
        class="loginButton"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
